cmd/juju/controller/destroy.go

```diff
@@ -165,8 +165,10 @@ func (c *destroyCommand) Run(ctx *cmd.Context) error {
 
 		// Even if we've not just requested for hosted models to be destroyed,
 		// there may be some being destroyed already. We need to wait for them.
+		// Check for both undead models and live machines, as machines may be
+		// in the controller model.
 		ctx.Infof("Waiting for hosted model resources to be reclaimed")
-		for ; hasUnDeadModels(modelsStatus); ctrStatus, modelsStatus = updateStatus(2 * time.Second) {
+		for ; hasUnDeadModels(modelsStatus) || ctrStatus.HostedMachineCount > 0; ctrStatus, modelsStatus = updateStatus(2 * time.Second) {
```

Comment:
With the loop getting this complicated (and sort of repeated twice), I think it warrants some refactoring.

Maybe `updateStatus` should return a single encapsulating object, which a (renamed) `hasUnDeadModels` should take and return a bool?